**BÀI TẬP CHƯƠNG 2 - HỌC PHẦN: Truy vấn SQL**

**26_Tran Duc Minh_23174600132_DHKL17A3HN**

- Cho hai bảng student(student_id, name, class, course_id, score) và course(id, course_name)
- Sử dụng jupyter notebook để chạy code, hiển thị kết quả thu được và đưa ra kết luận/câu trả lời về kết quả cuối cùng thu được. 

Kết nối đến SQLite và tạo bảng student và bảng course

In [3]:
import sqlite3
import pandas as pd

# Kết nối SQLite 
conn = sqlite3.connect("students.db") 
cursor = conn.cursor()

In [4]:
cursor.execute("DROP TABLE IF EXISTS student;")
cursor.execute("DROP TABLE IF EXISTS course;")

# Tạo bảng student
cursor.execute("""
CREATE TABLE student (
    student_id INTEGER PRIMARY KEY,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
);
""")

# Tạo bảng course
cursor.execute("""
CREATE TABLE course (
    id INTEGER PRIMARY KEY,
    course_name TEXT
);
""")

# Lưu thay đổi vào database
conn.commit()

In [5]:
# Dữ liệu cho bảng student
student_data = [
    (1, "Nguyen Minh Hoang", "May Tinh", 12, 6.7),
    (2, "Tran Thi Lan", "Kinh Te", 34, 9.2),
    (3, "Pham Van Nam", "Toan Tin", None, 7.9),   
    (4, "Le Thanh Huyen", "Toan Tin", 20, 7.2),
    (5, "Vu Quoc Anh", "May Tinh", 24, 8.0),
    (6, "Dang Thuy Linh", "May Tinh", 24, 5.5),
    (7, "Bui Tien Dung", "Kinh Te", 34, 9.2),
    (8, "Ho Ngoc Mai", "Toan Tin", 20, 8.8),
    (9, "Duong Huu Phuc", "Kinh Te", None, 7.2),   
    (10, "Cao Thi Hanh", "May Tinh", None, 7.0)     
]

# Dữ liệu cho bảng course
course_data = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]

# Chèn dữ liệu vào bảng student
cursor.executemany("INSERT INTO student VALUES (?, ?, ?, ?, ?);", student_data)

# Chèn dữ liệu vào bảng course
cursor.executemany("INSERT INTO course VALUES (?, ?);", course_data)

conn.commit()


In [6]:
# Đọc dữ liệu từ bảng student
student_df = pd.read_sql_query("SELECT * FROM student;", conn)
print("Bảng Student:")
display(student_df)

# Đọc dữ liệu từ bảng course
course_df = pd.read_sql_query("SELECT * FROM course;", conn)
print("Bảng Course:")
display(course_df)

Bảng Student:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7
1,2,Tran Thi Lan,Kinh Te,34.0,9.2
2,3,Pham Van Nam,Toan Tin,NaN,7.9
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2
4,5,Vu Quoc Anh,May Tinh,24.0,8.0
5,6,Dang Thuy Linh,May Tinh,24.0,5.5
6,7,Bui Tien Dung,Kinh Te,34.0,9.2
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2
9,10,Cao Thi Hanh,May Tinh,NaN,7.0


Bảng Course:


,id,course_name
0,12,Giai tich
1,26,Tin hoc
2,34,Thong ke


Nhận xét về bảng Student:

Bảng dữ liệu Student gồm 10 bản ghi với 5 trường thông tin. Cụ thể:

student_id: định danh của sinh viên, thuộc kiểu số nguyên
name: họ tên sinh viên, thuộc kiểu chuỗi
class: lớp học của sinh viên, thuộc kiểu chuỗi
course_id: mã môn học, thuộc kiểu số thực và có xuất hiện dữ liệu khuyết
score: điểm số của sinh viên, thuộc kiểu số thực

Qua quan sát, trường course_id tồn tại một số giá trị bị thiếu (NaN), cụ thể là 3 trường hợp tương ứng với các sinh viên có mã student_id là 3, 9 và 10. Điều này cho thấy dữ liệu chưa hoàn toàn đầy đủ và có thể cần xử lý trước khi phân tích.

Nhận xét về bảng Course:

Bảng Course bao gồm 3 dòng dữ liệu và 2 thuộc tính:

id: mã môn học, kiểu số nguyên
course_name: tên môn học, kiểu chuỗi

Dữ liệu trong bảng này đầy đủ, không có giá trị thiếu. Vì vậy, bảng Course có thể được sử dụng như một bảng tham chiếu để liên kết và bổ sung thông tin môn học cho bảng Student thông qua trường course_id.


**1. Hãy kết nối hai bảng trên theo những cách sau:**
**- Sử dụng tích Decartes.**
**- Sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN.**

Sử dụng tích Decartes

In [7]:
print("Tích Decartes:")
cursor.execute('''
SELECT * FROM student, course;
''')
df_decartes = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_decartes)
print("\n")


Tích Decartes:
    student_id               name     class  course_id  score  id course_name
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12   Giai tich
1            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  26     Tin hoc
2            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  34    Thong ke
3            2       Tran Thi Lan   Kinh Te       34.0    9.2  12   Giai tich
4            2       Tran Thi Lan   Kinh Te       34.0    9.2  26     Tin hoc
5            2       Tran Thi Lan   Kinh Te       34.0    9.2  34    Thong ke
6            3       Pham Van Nam  Toan Tin        NaN    7.9  12   Giai tich
7            3       Pham Van Nam  Toan Tin        NaN    7.9  26     Tin hoc
8            3       Pham Van Nam  Toan Tin        NaN    7.9  34    Thong ke
9            4     Le Thanh Huyen  Toan Tin       20.0    7.2  12   Giai tich
10           4     Le Thanh Huyen  Toan Tin       20.0    7.2  26     Tin hoc
11           4     Le Thanh Huyen  Toan Tin      

Kết luận:

Bảng dữ liệu thu được là kết quả của việc kết hợp toàn bộ giữa hai bảng Student và Course theo dạng tích Descartes. Do đó, mỗi sinh viên đều được ghép với tất cả các môn học có trong bảng Course, dẫn đến việc dữ liệu bị lặp lại theo từng tổ hợp.

Cụ thể, với 10 sinh viên và 3 môn học, bảng kết quả tạo ra 30 dòng dữ liệu, phù hợp với nguyên lý của phép tích Descartes (số dòng bằng tích số phần tử của hai bảng).

Việc xuất hiện các giá trị thiếu (NaN) trong trường course_id cho thấy vẫn tồn tại những sinh viên chưa được gán môn học cụ thể. Tuy nhiên, các bản ghi này vẫn xuất hiện trong kết quả do đặc điểm ghép toàn bộ của phép toán.

Bên cạnh đó, có thể nhận thấy rằng giá trị ở các cột như course_id và score chưa phản ánh chính xác mối liên hệ thực tế với bảng Course. Đồng thời, điểm số của mỗi sinh viên bị lặp lại ở nhiều môn học khác nhau, điều này cho thấy dữ liệu hiện tại chưa thể hiện đúng quan hệ giữa sinh viên và môn học.

Vì vậy, để đảm bảo tính chính xác và ý nghĩa trong phân tích, dữ liệu cần được xử lý lại bằng các phương pháp ghép phù hợp hơn (chẳng hạn như join theo khóa liên kết) thay vì sử dụng tích Descartes.


Sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN

INNER JOIN

In [8]:
print("INNER JOIN:")
cursor.execute('''
SELECT * FROM student INNER JOIN course ON student.course_id = course.id;
''')
df_inner = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_inner)
print("\n")

INNER JOIN:
   student_id               name     class  course_id  score  id course_name
0           1  Nguyen Minh Hoang  May Tinh         12    6.7  12   Giai tich
1           2       Tran Thi Lan   Kinh Te         34    9.2  34    Thong ke
2           7      Bui Tien Dung   Kinh Te         34    9.2  34    Thong ke




Kết luận:

Bảng kết quả chỉ bao gồm những bản ghi có sự tương ứng giữa trường course_id của bảng Student và trường id của bảng Course. Điều này cho thấy phép ghép chỉ giữ lại các dòng có giá trị phù hợp ở cả hai bảng.

Số lượng sinh viên xuất hiện trong bảng khá hạn chế, chỉ gồm 3 trường hợp. Điều này phản ánh rằng chỉ có một số ít sinh viên có thông tin môn học khớp với dữ liệu trong bảng Course, trong khi các trường hợp còn lại không thỏa điều kiện liên kết nên đã bị loại bỏ khỏi kết quả.

LEFT JOIN

In [9]:
print("LEFT JOIN:")
cursor.execute('''
SELECT * FROM student LEFT JOIN course ON student.course_id = course.id;
''')
df_left = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_left)
print("\n")

LEFT JOIN:
   student_id               name     class  course_id  score    id course_name
0           1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   Giai tich
1           2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0    Thong ke
2           3       Pham Van Nam  Toan Tin        NaN    7.9   NaN        None
3           4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN        None
4           5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN        None
5           6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN        None
6           7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0    Thong ke
7           8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN        None
8           9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN        None
9          10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN        None




Kết luận:

Phép LEFT JOIN đảm bảo giữ nguyên toàn bộ các bản ghi từ bảng Student, đồng thời chỉ bổ sung thông tin từ bảng Course trong trường hợp tồn tại giá trị liên kết phù hợp.

Những giá trị thiếu (NaN) xuất hiện tại các cột id và course_name cho thấy có không ít sinh viên có course_id không trùng khớp với bất kỳ mã môn học nào trong bảng Course. Điều này phản ánh sự không đồng nhất giữa hai bảng dữ liệu và cho thấy cần kiểm tra hoặc chuẩn hóa lại dữ liệu để đảm bảo tính liên kết chính xác.

RIGHT JOIN

In [10]:
print("RIGHT JOIN:")
cursor.execute('''
SELECT * FROM course LEFT JOIN student ON course.id = student.course_id;
''')
df_right = pd.DataFrame(cursor.fetchall(), columns=['id', 'course_name', 'student_id', 'name', 'class', 'course_id', 'score'])
print(df_right)
print("\n")

RIGHT JOIN:
   id course_name  student_id               name     class  course_id  score
0  12   Giai tich         1.0  Nguyen Minh Hoang  May Tinh       12.0    6.7
1  26     Tin hoc         NaN               None      None        NaN    NaN
2  34    Thong ke         7.0      Bui Tien Dung   Kinh Te       34.0    9.2
3  34    Thong ke         2.0       Tran Thi Lan   Kinh Te       34.0    9.2




Kết luận:

Phép RIGHT JOIN giữ lại toàn bộ các bản ghi từ bảng Course, đồng thời chỉ bổ sung thông tin từ bảng Student khi tồn tại sự liên kết phù hợp giữa hai bảng.

Quan sát kết quả cho thấy có trường hợp môn học "Tin học" xuất hiện nhưng các thông tin liên quan đến sinh viên đều bị thiếu (NaN). Điều này cho thấy chưa có sinh viên nào đăng ký môn học này, hoặc dữ liệu đăng ký hiện tại chưa được cập nhật đầy đủ.

FULL OUTER JOIN (mô phỏng bằng UNION của LEFT JOIN và RIGHT JOIN)

In [11]:
query = '''
SELECT * FROM student LEFT JOIN course 
ON student.course_id = course.id
UNION ALL
SELECT * FROM course LEFT JOIN student 
ON course.id = student.course_id
WHERE student.course_id IS NULL;
'''
df = pd.read_sql_query(query, conn)
print("FULL OUTER JOIN: ", df)


FULL OUTER JOIN:      student_id               name     class  course_id  score    id  \
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   
1            2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0   
2            3       Pham Van Nam  Toan Tin        NaN    7.9   NaN   
3            4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN   
4            5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN   
5            6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN   
6            7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0   
7            8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN   
8            9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN   
9           10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN   
10          26            Tin hoc      None        NaN    NaN   NaN   

   course_name  
0    Giai tich  
1     Thong ke  
2         None  
3         None  
4         None  
5         None  
6     Thon

Kết luận:

Phép FULL OUTER JOIN kết hợp toàn bộ dữ liệu từ cả hai bảng Student và Course, đồng thời sử dụng giá trị thiếu (NaN) cho những trường hợp không tồn tại sự liên kết tương ứng giữa hai bảng.

Trong kết quả, xuất hiện bản ghi với id = 26 và course_name = "Tin học" nhưng không có thông tin sinh viên đi kèm. Điều này cho thấy môn học này chưa có sinh viên đăng ký, hoặc dữ liệu liên quan đến việc đăng ký môn học vẫn chưa được cập nhật đầy đủ.

**2. Hãy cập nhật những giá trị course_id còn thiếu trong bảng student bằng câu lệnh SQL, trong đó các giá trị được điền là những giá trị nằm trong bảng course và loại bỏ những bản ghi tham gia những môn học không tồn tại bảng course. Sau đó hãy cho biết:**

**a. Tổng số sinh viên, điểm trung bình của từng lớp.**

**b. Tổng số sinh viên, điểm trung bình của từng môn học.**

**c. Phân loại thi đua theo số điểm của từng môn học biết:** 

**-   i. Điểm TB ≥  9.0: Xuất sắc.** 

**-   ii. 6.0 ≤ Điểm TB ≤ 8.9: Tốt.** 

**-  iii. Điểm TB < 6.0: Kém.** 

In [12]:

cursor.execute('''
UPDATE student
SET course_id = (
    SELECT id FROM course ORDER BY RANDOM() LIMIT 1
)
WHERE course_id IS NULL;
''')
conn.commit()


cursor.execute('''
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course);
''')
conn.commit()


cursor.execute('SELECT * FROM student;')
print(pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score']))


   student_id               name     class  course_id  score
0           1  Nguyen Minh Hoang  May Tinh         12    6.7
1           2       Tran Thi Lan   Kinh Te         34    9.2
2           3       Pham Van Nam  Toan Tin         26    7.9
3           7      Bui Tien Dung   Kinh Te         34    9.2
4           9     Duong Huu Phuc   Kinh Te         26    7.2
5          10       Cao Thi Hanh  May Tinh         26    7.0


Kết luận:

Các bản ghi trước đó có giá trị course_id bị thiếu (NaN) đã được thay thế bằng những mã môn học được chọn từ bảng Course.

Đồng thời, những trường hợp có course_id không tồn tại trong bảng Course cũng đã được loại bỏ khỏi dữ liệu.

Sau quá trình xử lý, bảng kết quả chỉ còn lại các sinh viên có course_id hợp lệ, đồng thời không còn xuất hiện giá trị thiếu, giúp dữ liệu trở nên nhất quán và sẵn sàng cho các bước phân tích tiếp theo.

a. Tổng số sinh viên, điểm trung bình của từng lớp.

In [13]:
# Truy vấn tổng số sinh viên và điểm trung bình của từng lớp
query = '''
SELECT class, COUNT(student_id) AS total_students, ROUND(AVG(score), 2) AS avg_score
FROM student
GROUP BY class;
'''

df = pd.read_sql_query(query, conn)
print(df)


      class  total_students  avg_score
0   Kinh Te               3       8.53
1  May Tinh               2       6.85
2  Toan Tin               1       7.90


Kết luận:

Kinh Te: 8.53 — Điểm trung bình cao nhất trong các lớp, chất lượng học tập tốt.

May Tinh: 6.85 — Điểm trung bình thấp nhất, cần cải thiện chất lượng học tập.

Toan Tin: 7.90 — Ở mức khá, có thể duy trì và cải thiện thêm. Lớp Toan Tin chỉ có 1 sinh viên, có thể ảnh hưởng đến tính đại diện của điểm trung bình.

b. Tổng số sinh viên, điểm trung bình của từng môn học. 

In [27]:
# Truy vấn tổng số sinh viên và điểm trung bình của từng môn học
query = '''
SELECT course.course_name, COUNT(student.student_id) AS total_students, ROUND(AVG(student.score), 2) AS avg_score
FROM student
LEFT JOIN course ON student.course_id = course.id
GROUP BY course.course_name;
'''

df = pd.read_sql_query(query, conn)
print(df)


  course_name  total_students  avg_score
0   Giai tich               1       6.70
1    Thong ke               2       9.20
2     Tin hoc               3       7.37


Kết luận:

Thong ke là môn có kết quả tốt nhất với điểm trung bình cao.

Giai tich có thể cần cải thiện về cả số lượng sinh viên đăng ký và chất lượng học tập.

Tin hoc nhiều sinh viên nhưng điểm chưa cao, cần tập trung cải thiện.

c. Phân loại thi đua theo số điểm của từng môn học biết: 

- i. Điểm TB ≥  9.0: Xuất sắc. 
- ii. 6.0 ≤ Điểm TB ≤ 8.9: Tốt. 
- iii. Điểm TB < 6.0: Kém.

In [15]:
# Truy vấn phân loại thi đua theo điểm trung bình của từng môn học
query = '''
SELECT course.course_name,
       ROUND(AVG(student.score), 2) AS avg_score,
       CASE
           WHEN ROUND(AVG(student.score), 2) >= 9.0 THEN 'Xuất sắc'
           WHEN ROUND(AVG(student.score), 2) BETWEEN 6.0 AND 8.9 THEN 'Tốt'
           ELSE 'Kém'
       END AS rating
FROM student
LEFT JOIN course ON student.course_id = course.id
GROUP BY course.course_name;
'''

df = pd.read_sql_query(query, conn)
print(df)


  course_name  avg_score    rating
0   Giai tich       6.70       Tốt
1    Thong ke       9.20  Xuất sắc
2     Tin hoc       7.37       Tốt


Kết luận:

Môn Thong ke dẫn đầu với xếp loại Xuất sắc và điểm trung bình cao nhất.

Giai tich và Tin hoc xếp loại Tốt, cần nâng cao chất lượng.

Không có môn Kém 

**3. Hãy xếp hạng sinh viên thông qua:** 

**a. Điểm số.** 

**b. Điểm số theo lớp học.** 

**c. Điểm số theo mã môn học.** 

**và cho biết top 3 sinh viện đạt thứ hạng cao nhất, top 3 sinh viên đạt thứ hạng thấp nhất theo 
từng trường hợp trên.**

a. Điểm số.

In [16]:
# Truy vấn xếp hạng sinh viên theo điểm số
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (ORDER BY score DESC) AS rank
FROM student
ORDER BY score DESC;
'''

df_rank = pd.read_sql_query(query, conn)
print(df_rank)


   student_id               name     class  course_id  score  rank
0           2       Tran Thi Lan   Kinh Te         34    9.2     1
1           7      Bui Tien Dung   Kinh Te         34    9.2     1
2           3       Pham Van Nam  Toan Tin         26    7.9     3
3           9     Duong Huu Phuc   Kinh Te         26    7.2     4
4          10       Cao Thi Hanh  May Tinh         26    7.0     5
5           1  Nguyen Minh Hoang  May Tinh         12    6.7     6


Kết luận:

Lớp Kinh Te: Đây là nhóm có thành tích áp đảo nhất. Hai vị trí dẫn đầu bảng (đồng hạng 1) đều thuộc về sinh viên lớp Kinh Te là Trần Thị Lan và Bùi Tiến Dũng với số điểm rất cao: 9.2. Sinh viên còn lại của lớp là Dương Hữu Phúc cũng đạt mức điểm khá là 7.2 (hạng 4).

Lớp Toan Tin: Chỉ có một đại diện xuất hiện trong danh sách này là Phạm Văn Nam. Sinh viên này có mức điểm tốt là 7.9, vươn lên vị trí top 3.

Lớp May Tinh: Hai sinh viên của lớp Máy Tính là Cao Thị Hạnh và Nguyễn Minh Hoàng đang tạm xếp cuối trong nhóm dữ liệu này với mức điểm lần lượt là 7.0 và 6.7 (hạng 5 và hạng 6).

b. Điểm số theo lớp học. 

In [17]:
# Truy vấn xếp hạng sinh viên theo điểm số trong từng lớp
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank
FROM student
ORDER BY class, class_rank;
'''

df_class_rank = pd.read_sql_query(query, conn)
print(df_class_rank)


   student_id               name     class  course_id  score  class_rank
0           2       Tran Thi Lan   Kinh Te         34    9.2           1
1           7      Bui Tien Dung   Kinh Te         34    9.2           1
2           9     Duong Huu Phuc   Kinh Te         26    7.2           3
3          10       Cao Thi Hanh  May Tinh         26    7.0           1
4           1  Nguyen Minh Hoang  May Tinh         12    6.7           2
5           3       Pham Van Nam  Toan Tin         26    7.9           1


Kết luận:

Lớp Kinh Te có sự cạnh tranh cao với khoảng cách điểm lớn giữa các hạng.

Lớp May Tinh cần nâng cao chất lượng học tập do điểm số dẫn đầu chưa thực sự nổi bật.

Lớp Toan Tin có mặt bằng điểm khá nhưng thiếu sự cạnh tranh.

c. Điểm số theo mã môn học. 

In [18]:
# Truy vấn xếp hạng sinh viên theo điểm số trong từng mã môn học
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank
FROM student
ORDER BY course_id, course_rank;
'''

df_course_rank = pd.read_sql_query(query, conn)
print(df_course_rank)


   student_id               name     class  course_id  score  course_rank
0           1  Nguyen Minh Hoang  May Tinh         12    6.7            1
1           3       Pham Van Nam  Toan Tin         26    7.9            1
2           9     Duong Huu Phuc   Kinh Te         26    7.2            2
3          10       Cao Thi Hanh  May Tinh         26    7.0            3
4           2       Tran Thi Lan   Kinh Te         34    9.2            1
5           7      Bui Tien Dung   Kinh Te         34    9.2            1


Kết luận:

Môn Giai tich: Nguyễn Minh Hoàng là sinh viên duy nhất với 6.7 điểm (mức trung bình).

Môn Tin hoc: Phạm Văn Nam dẫn đầu với 7.9 điểm; Dương Hữu Phúc và Cao Thị Hạnh lần lượt đạt 7.2 và 7.0 điểm. Điểm trung bình khá ổn.

Môn Thong ke: Trần Thị Lan và Bùi Tiến Dũng cùng đạt 9.2 điểm (xuất sắc), có điểm trung bình cao nhất.

Cho biết top 3 sinh viện đạt thứ hạng cao nhất, top 3 sinh viên đạt thứ hạng thấp nhất theo từng trường hợp trên.

Top 3 cao nhất & thấp nhất theo điểm số

In [19]:
# Truy vấn top 3 cao nhất và thấp nhất theo điểm số toàn bộ
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (ORDER BY score DESC) AS rank_desc,
           RANK() OVER (ORDER BY score ASC) AS rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE rank_desc <= 3 OR rank_asc <= 3
ORDER BY rank_desc;
'''
df_top_all = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:")
print(df_top_all)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:
   student_id               name     class  course_id  score  rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2          1   
1           7      Bui Tien Dung   Kinh Te         34    9.2          1   
2           3       Pham Van Nam  Toan Tin         26    7.9          3   
3           9     Duong Huu Phuc   Kinh Te         26    7.2          4   
4          10       Cao Thi Hanh  May Tinh         26    7.0          5   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7          6   

   rank_asc  
0         5  
1         5  
2         4  
3         3  
4         2  
5         1  




Top 3 cao nhất:

Trần Thị Lan và Bùi Tiến Dũng (9.2 điểm) — Hạng 1.

Phạm Văn Nam (7.9 điểm) — Hạng 3.

Top 3 thấp nhất:

Nguyễn Minh Hoàng (6.7 điểm) — Hạng 6.

Cao Thị Hạnh (7.0 điểm) — Hạng 5.

Dương Hữu Phúc (7.2 điểm) — Hạng 4.

Top 3 cao nhất & thấp nhất theo điểm số từng lớp

In [20]:
# Truy vấn top 3 cao nhất và thấp nhất theo điểm từng lớp
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank_desc,
           RANK() OVER (PARTITION BY class ORDER BY score ASC) AS class_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE class_rank_desc <= 3 OR class_rank_asc <= 3
ORDER BY class, class_rank_desc;
'''

df_top_class = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng lớp:")
print(df_top_class)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số từng lớp:
   student_id               name     class  course_id  score  class_rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2                1   
1           7      Bui Tien Dung   Kinh Te         34    9.2                1   
2           9     Duong Huu Phuc   Kinh Te         26    7.2                3   
3          10       Cao Thi Hanh  May Tinh         26    7.0                1   
4           1  Nguyen Minh Hoang  May Tinh         12    6.7                2   
5           3       Pham Van Nam  Toan Tin         26    7.9                1   

   class_rank_asc  
0               2  
1               2  
2               1  
3               2  
4               1  
5               1  




 Top 3 cao nhất & thấp nhất theo điểm số từng lớp
 
Lớp Kinh Tế:

Cao nhất: Trần Thị Lan và Bùi Tiến Dũng (9.2 điểm) — Đồng hạng 1.

Thấp nhất: Dương Hữu Phúc (7.2 điểm) — Hạng 3.

Lớp Máy Tính:

Cao nhất: Cao Thị Hạnh (7.0 điểm) — Hạng 1.

Thấp nhất: Nguyễn Minh Hoàng (6.7 điểm) — Hạng 2.

Lớp Toán Tin:

Cao nhất: Phạm Văn Nam (7.9 điểm) — Hạng 1.

Thấp nhất: Cũng là Phạm Văn Nam (7.9 điểm) — Hạng 1 (vì chỉ có 1 sinh viên).

Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học

In [28]:
# Truy vấn top 3 cao nhất và thấp nhất theo điểm từng môn học
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank_desc,
           RANK() OVER (PARTITION BY course_id ORDER BY score ASC) AS course_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE course_rank_desc <= 3 OR course_rank_asc <= 3
ORDER BY course_id, course_rank_desc;
'''

df_top_course = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học:")
print(df_top_course)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học:
   student_id               name     class  course_id  score  \
0           1  Nguyen Minh Hoang  May Tinh         12    6.7   
1           3       Pham Van Nam  Toan Tin         26    7.9   
2           9     Duong Huu Phuc   Kinh Te         26    7.2   
3          10       Cao Thi Hanh  May Tinh         26    7.0   
4           2       Tran Thi Lan   Kinh Te         34    9.2   
5           7      Bui Tien Dung   Kinh Te         34    9.2   

   course_rank_desc  course_rank_asc  
0                 1                1  
1                 1                3  
2                 2                2  
3                 3                1  
4                 1                1  
5                 1                1  




Top 3 cao nhất & thấp nhất theo điểm số từng mã môn học

Môn có course_id = 12

Chỉ có 1 sinh viên: Nguyễn Minh Hoàng (6.7) nên vừa là cao nhất vừa là thấp nhất

Môn có course_id = 26

Gồm các sinh viên:

Phạm Văn Nam: 7.9 (cao nhất)

Dương Hữu Phúc: 7.2

Cao Thị Hạnh: 7.0 (thấp nhất)

Môn có course_id = 34
Gồm các sinh viên:

Trần Thị Lan: 9.2

Bùi Tiến Dũng: 9.2

Cả hai sinh viên đều đạt điểm cao nhất (9.2)

**4. Hãy bổ sung thêm một trường graduation_date có kiểu dữ liệu là DATETIME vào bảng student để xác định thời gian tốt nghiệp của từng bạn, trong đó thời gian tốt nghiệp được xác định bởi thời gian hiện tại cộng với số hạng tương ứng của bạn đó tính theo điểm số.**

Thêm cột graduation_date vào bảng student

In [ ]:
# Thêm cột graduation_date với kiểu DATETIME 
cursor.execute('''
ALTER TABLE student ADD COLUMN graduation_date DATETIME;
''')
conn.commit()

Cập nhật graduation_date dựa trên thứ hạng

In [23]:
# Cập nhật graduation_date dựa trên thứ hạng
cursor.execute('''
WITH ranked_students AS (
    SELECT student_id, RANK() OVER (ORDER BY score DESC) AS rank
    FROM student
)
UPDATE student
SET graduation_date = DATETIME('now', '+' || (
    SELECT rank FROM ranked_students WHERE ranked_students.student_id = student.student_id
) || ' months');
''')
conn.commit()


 Kiểm tra lại bảng student với trường graduation_date

In [24]:
# Truy vấn lại để kiểm tra kết quả
query = '''
SELECT student_id, name, class, course_id, score, graduation_date
FROM student
ORDER BY score DESC;
'''

df_updated = pd.read_sql_query(query, conn)
print(" Bảng `student` sau khi thêm trường `graduation_date`:")
print(df_updated)


 Bảng `student` sau khi thêm trường `graduation_date`:
   student_id               name     class  course_id  score  \
0           2       Tran Thi Lan   Kinh Te         34    9.2   
1           7      Bui Tien Dung   Kinh Te         34    9.2   
2           3       Pham Van Nam  Toan Tin         26    7.9   
3           9     Duong Huu Phuc   Kinh Te         26    7.2   
4          10       Cao Thi Hanh  May Tinh         26    7.0   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7   

       graduation_date  
0  2026-05-02 07:05:08  
1  2026-05-02 07:05:08  
2  2026-07-02 07:05:08  
3  2026-08-02 07:05:08  
4  2026-09-02 07:05:08  
5  2026-10-02 07:05:08  


Bảng student sau khi thêm trường graduation_date:

Trần Thị Lan và Bùi Tiến Dũng (Kinh Tế):
- Điểm: 9.2 — Ngày tốt nghiệp: 02/05/2026.

Phạm Văn Nam (Toán Tin):
- Điểm: 7.9 — Ngày tốt nghiệp: 02/07/2026.

Dương Hữu Phúc (Kinh Tế):
- Điểm: 7.2 — Ngày tốt nghiệp: 02/08/2026.

Cao Thị Hạnh (Máy Tính):
- Điểm: 7.0 — Ngày tốt nghiệp: 02/09/2026.

Nguyễn Minh Hoàng (Máy Tính):
- Điểm: 6.7 — Ngày tốt nghiệp: 02/10/2026.

